### PR TITLE
Optimized all firebase calls to use tanstack query (react query)

### PR DESCRIPTION
**Pull request by FrankreedX**
Tanstack query automatically caches all requests so app's performance is drastically improved. Code in each pages are also massively simplified, with most fetching logic moved to hooks.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,12 +8,14 @@
     "ios": "expo start --ios",
     "web": "expo start --web",
     "pretty": "prettier --write .",
-    "python": "./venv/bin/python"
+    "python": "./venv/bin/python",
+    "firestore": "firebase emulators:start --only firestore"
```
Comment (solderq35) [inline package.json:12]: I connected to Firestore emulator (installed firebase-tools globally and edited firebaseConfig.js to connect), but I got an error (below).

Hess's email about this topic also said something about needing paid plan for Firebase Export (or something like that).

Is it better to remove all the firestore emulator stuff from this PR for now, or should we wait to see if Hess can upgrade the Firebase permissions?

image

File: firestore-debug.log [hunk 2]
```diff
@@ -0,0 +1,23 @@
+Mar 02, 2024 8:26:49 PM com.google.cloud.datastore.emulator.firestore.websocket.WebSocketServer start
```
Comment (solderq35) [inline firestore-debug.log:1]: regardless of if we keep firestore emulator, this should be in gitignore right?

**Comment by solderq35** [comment]
How does the cache work? After loading the team index page, I tried editing a user's name as a test. Clicking on the user showed the updated name on the next page (team > user profile), but the team index still showed old page.

Should we add option to manually refresh cache, like by pulling down page and triggering loading spinner? Maybe for future PR?
- refetchQueries (refresh cache?), invalidateQueries (refresh cache?), staleTime (max cache age?) all look promising from https://tanstack.com/query/v4/docs/reference/QueryClient

image

image

File: package.json (hunk 1 above)
Comment (FrankreedX) [reply package.json:12]: yes let's remove the emulator stuff

File: firestore-debug.log (hunk 2 above)
Comment (FrankreedX) [reply firestore-debug.log:1]: yep

**Comment by FrankreedX** [comment]
> How does the cache work? After loading the team index page, I tried editing a user's name as a test. Clicking on the user showed the updated name on the next page (team > user profile), but the team index still showed old page.
> 
> Should we add option to manually refresh cache, like by pulling down page and triggering loading spinner? Maybe for future PR?
> 
> * refetchQueries (refresh cache?), invalidateQueries (refresh cache?), staleTime (max cache age?) all look promising from https://tanstack.com/query/v4/docs/reference/QueryClient
> 
> image
> 
> image

Did the team -> user profile page load fast or did it take its time? I think the cache reload could've been triggered by you navigating to that page. The team page is not reloaded because it's always loaded due to the nav bar. That'll be my guess

**Comment by solderq35** [comment]
> Did the team -> user profile page load fast or did it take its time? I think the cache reload could've been triggered by you navigating to that page. The team page is not reloaded because it's always loaded due to the nav bar. That'll be my guess

I think at the time I hadn't navigated to any "team > user profile" page yet, so the cache for it wasn't set yet. My point was about the team index page I guess.

**Comment by solderq35** [comment]
> Did the team -> user profile page load fast or did it take its time? I think the cache reload could've been triggered by you navigating to that page. The team page is not reloaded because it's always loaded due to the nav bar. That'll be my guess

EDIT: I tried again, this time I loaded both team page and visited profile page of specific user before I edited the user's name in Firestore console.

Going from "team > user profile" to "team > user profile > stats" and then back to "team > user profile" didn't refresh the cache

Going back to "team index page", then going to "team > user profile" did refresh the cache.

**Comment by FrankreedX** [comment]
Ye, going back doesn't reload the page, only forwards. This is due to how the stack navigation works. I don't think our application is too time sensitive but we can indeed put in a pull to refresh and a timer. There should already be a default cache timeout i think

**Pull request merged**